src/utils.c

```diff
@@ -110,30 +117,221 @@ SEXP colnamesInt(SEXP x, SEXP cols, SEXP check_dups) {
+  if (isLanguage(expr) && CAR(expr)==sym_patterns && mode_sd) {
+    SEXP xnames = getAttrib(x, R_NamesSymbol);
+    if (isNull(xnames))
+      error(_("'x' argument data.table has no names"));
+    SEXP sym_quote = install("quote"); // .SDcols = Reduce(intersect, do_patterns(colsub, names_x))
+    SEXP sym_do_patterns = install("do_patterns");
+    SEXP lpat = PROTECT(eval(LCONS(sym_do_patterns, LCONS(LCONS(sym_quote, LCONS(expr, R_NilValue)), LCONS(xnames, R_NilValue))), rho)); protecti++;
+    SEXP sym_Reduce = install("Reduce");
+    SEXP sym_intersect = install("intersect");
+    //cols = PROTECT(eval(LCONS(sym_Reduce, LCONS(sym_intersect, LCONS(lpat, R_NilValue))), rho)); protecti++;
+    UNPROTECT(protecti);
+    return eval(LCONS(sym_Reduce, LCONS(sym_intersect, LCONS(lpat, R_NilValue))), rho);
+  }
+  // single symbol V2 might be also a function, to check that we need to evaluate, thus first we check if "V2" is existing column name, if not then we evaluate to see if it is a function.
+  // adding support of 'with' argument here may improve control, when with=FALSE then cols=eval(expr, rho) could be made straight away
+  SEXP expr_sym = R_NilValue, expr_sym_match = R_NilValue;
+  if (isSymbol(expr) && mode_j) {  // V1, is.numeric
+    SEXP xnames = getAttrib(x, R_NamesSymbol);
+    if (isNull(xnames))
+      error(_("'x' argument data.table has no names"));
+    expr_sym = PROTECT(ScalarString(PRINTNAME(expr))); protecti++; // "V1", "is.numeric"
+    expr_sym_match = chmatch(expr_sym, xnames, 0);
+    if (INTEGER(expr_sym_match)[0]!=0) {
+      UNPROTECT(protecti);
+      return expr_sym_match;
+    }
+  }
+  // various cases where all have to be evaluated: c("V1","V2"), paste0("V",1:2), is.numeric, function(x) is.numeric(x), also `cols` symbol when there is no column named "cols"
+  SEXP value = PROTECT(eval(expr, rho)); protecti++;
+  SEXP cols = R_NilValue;
+  if (isFunction(value) && mode_sd) { // expr could be either symbol or language: f, function(x) x
+    cols = PROTECT(funCols(x, expr, rho)); protecti++;
+  } else if (isSymbol(expr) && mode_j) { // handle error message consistently
+    if (isNull(expr_sym) || isNull(expr_sym_match))
+      error(_("internal error: expr symbol name, and if it matches to column names, should be known already")); // # nocov
+    cols = expr_sym;
+  } else if (isFunction(value) && mode_j) {
+    error(_("providing function as a column selection to 'j' argument is not supported, use '.SDcols' instead"));
+  } else { // evaluated value: c(2L,3L), c("V1","V2"), but not call objects anymore
+    cols = value;
+  }
+  if (isNull(cols))
+    error("internal error: cols should not be NULL by now"); // # or could be NULL when .SD provided, but eval(.SD) may already resolve that, no nocov for now
+  UNPROTECT(protecti);
+  return colnamesInt(x, cols, ScalarLogical(false), ScalarLogical(inverse));
```

Comment:
Should consider returning a named vector. We still need to get the variable names in ```[.data.table``` and since some of these expressions evaluate to character vectors, we would be doing the same work twice if we only return the integer locations.

Reply:
Names can be easily looked up `names(x)[ans]` so that it the last thing we should be worried about. Returning named vector will adds some extra overhead which might not always be needed.

Reply:
Regarding "overhead which may not always be needed", could you expand? I am on board with leaving it to ```names(x)[ans]``` but it seems like we would always need the vars names if users are selecting the columns this way for ```j```, ```by``` and ```.SDcols```.

Reply:
Generating character vector is much more complex in C. I would say it will be best to come back to your request when plugging this function into `[.data.table`, for now the aim is to have all rules in place.